llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp

```diff
@@ -0,0 +1,251 @@
+      if (!getInvokeeIfInvokeSimdCall(CI)) {
+        llvm_unreachable("unsupported data flow pattern for invoke_simd 12");
+      }
+      continue;
+    }
+    if (const auto *LI = dyn_cast<LoadInst>(V)) {
+      // LoadInst from this addr is OK, as it does not affect what can be stored
+      // through the addr
+      continue;
+    }
+    llvm_unreachable("unsupported data flow pattern for invoke_simd 13");
+  }
+}
+
+// Example1 (function is direct argument to _Z21__builtin_invoke_simd):
+// %call6.i = call spir_func float @_Z21__builtin_invoke_simd...(
+//   <16 x float> (float addrspace(4)*, <16 x float>, i32)* %28, <== function
+//   pointer float addrspace(4)* %arg1, float %arg2, i32 %arg3)
+//
+// Example 2 (invoke_simd's target function pointer flows through IR):
+// %"fptr_t = <16 x float> (float addrspace(4)*, <16 x float>, i32)*
+// ...
+// %fa_as0 = alloca %fptr_t
+// ...
+// %fa = addrspacecast %fptr_t* %fa_as0 to %fptr_t addrspace(4)*
+// ...
+// store %fptr_t @__SIMD_CALLEE, %fptr_t addrspace(4)* %fa
+// ...
+// %f = load %fptr_t, %fptr_t addrspace(4)* %fa
+// ...
+// %res = call spir_func float @_Z21__builtin_invoke_simd...(
+//   %fptr_t %f, <== function pointer
+//  float addrspace(4)* %arg1,
+//  float %arg2,
+//  i32 %arg3)
+//
+bool processInvokeSimdCall(CallInst *CI) {
+  Value *V = getInvokeeIfInvokeSimdCall(CI);
+
+  if (!V) {
+    llvm_unreachable(("bad use of " + Twine(INVOKE_SIMD_PREF)).str().c_str());
+  }
+  Function *SimdF = nullptr;
+  bool Modified = false;
+
+  if (!(SimdF = dyn_cast<Function>(V))) {
+    LoadInst *LI = dyn_cast<LoadInst>(stripCasts(V));
+
+    if (!LI) {
+      llvm_unreachable("unsupported data flow pattern for invoke_simd 0");
+    }
+    ValueSet Vals;
+    getPossibleStoredVals(LI->getPointerOperand(), Vals);
+
+    if (Vals.size() != 1 || !(SimdF = dyn_cast<Function>(*Vals.begin()))) {
+      llvm_unreachable("unsupported data flow pattern for invoke_simd 1");
+    }
+    // _Z21__builtin_invoke_simd invokee is an SSA value, replace it with the
+    // link-time constant SimdF as computed by getPossibleStoredVals
+    CallInst *CI1 = cast<CallInst>(CI->clone());
+    constexpr int SimdInvokeInvokeeArgIndex = 0;
+    CI1->setOperand(SimdInvokeInvokeeArgIndex, SimdF);
+    CI1->insertAfter(CI);
+    CI->replaceAllUsesWith(CI1);
+    CI->eraseFromParent();
+    Modified = true;
+  }
+  if (!SimdF->hasFnAttribute(llvm::genx::VCFunctionMD::VCStackCall)) {
+    SimdF->addFnAttr(llvm::genx::VCFunctionMD::VCStackCall);
+  }
+  return Modified;
+}
+} // namespace
+
+namespace llvm {
+PreservedAnalyses SYCLLowerInvokeSimdPass::run(Module &M,
+                                               ModuleAnalysisManager &MAM) {
+  bool Modified = false;
+
+  for (Function &F : M) {
+    if (!F.isDeclaration() || !F.getName().startswith(INVOKE_SIMD_PREF)) {
+      continue;
+    }
+    SmallVector<User *, 4> Users(F.users().begin(), F.users().end());
```

Comment:
```suggestion
    SmallVector<User *, 4> Users(F.users());
```

Theoretically, that should work: `SmallVector` accepts `iterator_range` in one of its constructors and `Value::users()` returns exactly it.